quint/src/itf.ts

```diff
@@ -126,6 +173,60 @@ export function toItf(vars: string[], states: QuintEx[]): Either<string, ItfTrac
   })
 }
 
-// export function ofItf(itf: any[]): Either<string, QuintEx[]> {
-//   const stateToExpr = (any: )
-// }
+export function ofItf(itf: ItfTrace): QuintEx[] {
+  // Benign state to synthesize ids for the Quint expressions
+  var nextId = 0n
+  // Produce the next ID in sequence
+  const getId = (): bigint => {
+    const id = nextId
+    nextId = nextId + 1n
+    return id
+  }
+
+  const ofItfValue = (value: ItfValue): QuintEx => {
+    const withId = { id: getId() }
+    if (typeof value === 'boolean') {
+      return { ...withId, kind: 'bool', value }
+    } else if (typeof value === 'string') {
+      return { ...withId, kind: 'str', value }
+    } else if (typeof value === 'number') {
+      return { ...withId, kind: 'int', value: BigInt(value) }
+    } else if (Array.isArray(value)) {
+      return { ...withId, kind: 'app', opcode: 'List', args: value.map(ofItfValue) }
+    } else if (isBigint(value)) {
+      return { ...withId, kind: 'int', value: BigInt(value['#bigint']) }
+    } else if (isTup(value)) {
+      return { ...withId, kind: 'app', opcode: 'Tup', args: value['#tup'].map(ofItfValue) }
+    } else if (isSet(value)) {
+      return { ...withId, kind: 'app', opcode: 'Set', args: value['#set'].map(ofItfValue) }
+    } else if (isUnserializable(value)) {
+      return { ...withId, kind: 'name', name: value['#unserializable'] }
+    } else if (isMap(value)) {
+      const args = value['#map'].map(([key, value]) => {
+        const k = ofItfValue(key)
+        const v = ofItfValue(value)
+        return { id: getId(), kind: 'app', opcode: 'Tup', args: [k, v] } as QuintApp
+      })
+      return {
+        ...withId,
+        kind: 'app',
+        opcode: 'Map',
+        args,
+      }
+    } else if (typeof value === 'object') {
+      // Any other object must represent a record
+      // For each key/value pair in the object, form the quint expressions representing
+      // the record field and value
+      const args = Object.keys(value)
+        .filter(key => key !== '#meta') // Must be removed from top-level ojects representing states
+        .map(f => [{ id: getId(), kind: 'str', value: f }, ofItfValue(value[f])] as [QuintStr, QuintEx])
+        .flat() // flatten the converted pairs of fields into a single array
+      return { ...withId, kind: 'app', opcode: 'Rec', args }
```

Comment:
Very minor aesthetic thing: you can do this
```suggestion
    const id = getId()
    if (typeof value === 'boolean') {
      return { id, kind: 'bool', value }
    } else if (typeof value === 'string') {
      return { id, kind: 'str', value }
    } else if (typeof value === 'number') {
      return { id, kind: 'int', value: BigInt(value) }
    } else if (Array.isArray(value)) {
      return { id, kind: 'app', opcode: 'List', args: value.map(ofItfValue) }
    } else if (isBigint(value)) {
      return { id, kind: 'int', value: BigInt(value['#bigint']) }
    } else if (isTup(value)) {
      return { id, kind: 'app', opcode: 'Tup', args: value['#tup'].map(ofItfValue) }
    } else if (isSet(value)) {
      return { id, kind: 'app', opcode: 'Set', args: value['#set'].map(ofItfValue) }
    } else if (isUnserializable(value)) {
      return { id, kind: 'name', name: value['#unserializable'] }
    } else if (isMap(value)) {
      const args = value['#map'].map(([key, value]) => {
        const k = ofItfValue(key)
        const v = ofItfValue(value)
        return { id: getId(), kind: 'app', opcode: 'Tup', args: [k, v] } as QuintApp
      })
      return {
        id,
        kind: 'app',
        opcode: 'Map',
        args,
      }
    } else if (typeof value === 'object') {
      // Any other object must represent a record
      // For each key/value pair in the object, form the quint expressions representing
      // the record field and value
      const args = Object.keys(value)
        .filter(key => key !== '#meta') // Must be removed from top-level ojects representing states
        .map(f => [{ id: getId(), kind: 'str', value: f }, ofItfValue(value[f])] as [QuintStr, QuintEx])
        .flat() // flatten the converted pairs of fields into a single array
      return { id, kind: 'app', opcode: 'Rec', args }
```

Reply:
That is nicer. Thanks :)